packages/react-dom/src/client/DOMPropertyOperations.js

```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
   if (shouldIgnoreAttribute(name, propertyInfo, isCustomComponentTag)) {
     return;
   }
+
+  if (
+    enableCustomElementPropertySupport &&
+    isCustomComponentTag &&
+    name[0] === 'o' &&
+    name[1] === 'n'
+  ) {
+    let eventName = name.replace(/Capture$/, '');
+    const useCapture = name !== eventName;
+    const nameLower = eventName.toLowerCase();
+    if (nameLower in node) {
+      eventName = nameLower;
+    }
+    eventName = eventName.slice(2);
+
+    const listenersObjName = eventName + (useCapture ? 'true' : 'false');
+    const listeners = getCustomElementEventHandlersFromNode(node);
+    const alreadyHadListener = listeners[listenersObjName];
+
+    if (typeof value === 'function' || alreadyHadListener) {
```

Comment:
Hmm. So we do different things depending on the value type. This makes me a bit uneasy. I guess that's existing practice in Preact etc?

This makes me uneasy because props can always change midway. E.g. you get function on first render and something else on the second render. Do we have tests demonstrating what exactly happens when the type changes? The guarantee we want to preserve is that A -> B -> C -> D should have the same "end result" as just D, regardless of what A, B, C were. E.g. number -> string -> function -> number, or number -> function -> function -> string.

Reply:
Here's a concrete example.

This doesn't throw:

```js
      const container = document.createElement('div');
      // ReactDOM.render(<Test handler={oncustomevent} />, container);
      ReactDOM.render(<Test handler="hello" />, container);
      const customElement = container.querySelector('my-custom-element');
      customElement.dispatchEvent(new Event('customevent'));
```

but this throws:

```js
      const container = document.createElement('div');
      ReactDOM.render(<Test handler={oncustomevent} />, container);
      ReactDOM.render(<Test handler="hello" />, container);
      const customElement = container.querySelector('my-custom-element');
      customElement.dispatchEvent(new Event('customevent'));
```

>TypeError: (0 , _ReactDOMComponentTree.getCustomElementEventHandlersFromNode)(...)[(e.type + "false")] is not a function

This doesn't seem like consistent behavior.

Reply:
Thanks for pointing out this case! I like the fully consistent behavior option.

More on the function type behavior - Preact doesn't look at the type of the value passed in and unconditionally forwards it to addEventListener. Jason Miller endorsed this behavior by telling me that it supports the [EventListener interface](https://developer.mozilla.org/en-US/docs/Web/API/EventListener) better because addEventListener can take objects in addition to just functions, and that nobody has ever complained about all properties with `on` being reserved.

I know that Sebastian seemed objected to reserving everything that starts with `on` (meaning you can't do `one={'foo'}` for example). I believe that this behavior of looking for `function`s strongly mitigates this issue, and I don't really see the EventListener interface being a big issue either. However, I don't exactly remember what Sebastian's last thoughts about this were, and I don't remember exactly when I decided to start looking for the `function` type. I did propose the behavior [in this comment](https://github.com/facebook/react/issues/11347#issuecomment-887889479), and there wasn't really anyone disagreeing.

In the end, I think that we should go forward with this and do the Fully consistent behavior option. I'll plan on implementing it once I get through the rest of the comments in this PR.

Reply:
I added some code here to implement the fully consistent option - when calling addEventListener, it will now try to figure out if we previously assigned the same prop as an attribute or property, in which case it will assign null into it.